File: patterns/prototype.js
//prototype pattern
//create copies from examples
//
// Define the Prototype interface, which specifies the clone method
class Prototype {
    clone() { }
}

// Define Concrete Prototypes, which implement the Prototype interface and provide a way to clone themselves
class ConcretePrototype1 extends Prototype {
    constructor() {
        super();
        this.property1 = "Property 1";
        this.property2 = "Property 2";
    }

    clone() {
        return new ConcretePrototype1();
    }
}

class ConcretePrototype2 extends Prototype {
    constructor() {
        super();
        this.property3 = "Property 3";
        this.property4 = "Property 4";
    }

    clone() {
        return new ConcretePrototype2();
    }
}

// Use the Concrete Prototypes to create new objects
const prototype1 = new ConcretePrototype1();
const clone1 = prototype1.clone();

const prototype2 = new ConcretePrototype2();
const clone2 = prototype2.clone();

File: patterns/prototype.py
import copy

# Define the Prototype class, which is the class to be cloned
class Prototype:
    def clone(self):
        pass

# Define the ConcretePrototype classes, which are the classes to be cloned
class ConcretePrototypeA(Prototype):
    def __init__(self, value):
        self._value = value

    def get_value(self):
        return self._value

    def clone(self):
        return copy.copy(self)

class ConcretePrototypeB(Prototype):
    def __init__(self, value):
        self._value = value

    def get_value(self):
        return self._value

    def clone(self):
        return copy.copy(self)

# Use the Prototype pattern to clone objects
prototype_a = ConcretePrototypeA(1)
prototype_b = ConcretePrototypeB(2)

clone_a = prototype_a.clone()
clone_b = prototype_b.clone()

print(clone_a.get_value())
print(clone_b.get_value())
